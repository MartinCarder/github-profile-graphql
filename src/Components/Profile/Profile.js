import React from 'react'
import PropTypes from 'prop-types'
import { ProfileImg } from './Profile.styles.js'

const Profile = ({ avatarUrl, bio, company, location, name, login }) => {
  return (
    <div>
      <ProfileImg img={avatarUrl} />
      <p data-id="name">{name}</p>
      <p data-id="login">{login}</p>
      {bio && <p data-id="bio">{bio}</p>}
      {company && <p data-id="company">{company}</p>}
      {location && <p data-id="location">{location}</p>}
    </div>
  )
}

Profile.propTypes = {
  avatarUrl: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  login: PropTypes.string.isRequired,
  bio: PropTypes.string,
  company: PropTypes.string,
  location: PropTypes.string,
}

export default Profile
