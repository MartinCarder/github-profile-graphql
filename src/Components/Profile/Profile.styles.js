import styled from 'styled-components'

export const ProfileImg = styled.div`
  width: 230px;
  height: 230px;

  background: url('${props => props.img}') no-repeat center center;
  background-size: cover;
  border-radius: 6px;
  overflow: hidden;
  margin: 0 auto;
`
