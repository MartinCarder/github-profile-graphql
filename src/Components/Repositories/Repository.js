import React from 'react'
import { Container } from './Repository.styles'
import PropTypes from 'prop-types'

const Repository = ({ name, description }) => {
  return (
    <Container>
      <p>{name}</p>
      <p>{description}</p>
    </Container>
  )
}

Repository.propTypes = {
  name: PropTypes.string.isRequired,
  description: PropTypes.string,
}

export default Repository
