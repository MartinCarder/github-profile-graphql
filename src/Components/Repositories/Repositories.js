import React from 'react'
import Repository from './Repository'
import PropTypes from 'prop-types'

const Repositories = ({ repositoriesList }) => {
  return <div>{repositoriesList.nodes.map(item => <Repository key={item.id} {...item} />)}</div>
}

Repositories.propTypes = {
  repositoriesList: PropTypes.shape({
    nodes: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
      })
    ).isRequired,
  }),
}

export default Repositories
