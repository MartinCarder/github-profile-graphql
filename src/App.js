import React, { Component } from 'react'
import gql from 'graphql-tag'
import { Query } from 'react-apollo'
import PropTypes from 'prop-types'
import Profile from './Components/Profile/Profile'
import Repositories from './Components/Repositories/Repositories'

export const GET_USER = gql`
  {
    user(login: "gaearon") {
      avatarUrl
      bio
      company
      name
      location
      login
      repositories(first: 50, orderBy: { field: PUSHED_AT, direction: DESC }) {
        pageInfo {
          endCursor
          startCursor
        }
        nodes {
          id
          name
          description
        }
      }
    }
  }
`

export const ComponentWrapper = ({ loading, error, data }) => {
  if (loading) return 'Loading...'
  if (error) return `Error! ${error.message}`
  const { user } = data
  return (
    <div className="App">
      <Profile {...user} />
      <Repositories repositoriesList={user.repositories} />
    </div>
  )
}

ComponentWrapper.propTypes = {
  loading: PropTypes.bool.isRequired,
  error: PropTypes.oneOfType([
    PropTypes.shape({
      message: PropTypes.string.isRequired,
    }),
    PropTypes.bool,
  ]),
  data: PropTypes.shape({
    user: PropTypes.shape({
      repositories: PropTypes.object.isRequired,
    }),
  }),
}

class App extends Component {
  render() {
    return <Query query={GET_USER}>{props => <ComponentWrapper {...props} />}</Query>
  }
}

export default App
