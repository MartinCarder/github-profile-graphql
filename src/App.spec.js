import React from 'react'
import { ApolloProvider } from 'react-apollo'
import { ApolloClient } from 'apollo-client'
import { createHttpLink } from 'apollo-link-http'
import { InMemoryCache } from 'apollo-cache-inmemory'
import createMockedNetworkFetch from 'apollo-mocknetworkinterface'
import { mount, shallow } from 'enzyme'
import App, { GET_USER, ComponentWrapper } from './App'
import { Query } from 'react-apollo'
import Profile from './Components/Profile/Profile'
import Repositories from './Components/Repositories/Repositories'

const createResponse = () => {
  return {
    data: {
      user: {
        id: 1,
        avatarUrl: '',
        bio: '',
        company: '',
        name: '',
        location: '',
        login: '',
        repositories: {
          pageInfo: {
            endCursor: '',
            startCursor: '',
          },
          nodes: [],
        },
      },
    },
  }
}

const mockedNetworkFetch = createMockedNetworkFetch(createResponse, { timeout: 100 })
const client = new ApolloClient({
  link: createHttpLink({ uri: 'http://localhost:3000', fetch: mockedNetworkFetch }),
  cache: new InMemoryCache({
    addTypename: false,
  }),
})

describe('<App />', () => {
  const data = {
    user: {
      repositories: { nodes: [] },
      name: 'martin carder',
      avatarUrl: 'http://',
      login: 'martincarder',
    },
  }

  it('should be the correct query', () => {
    expect(GET_USER).toMatchSnapshot()
  })

  it('snapshot of App', () => {
    const wrapper = mount(
      <ApolloProvider client={client}>
        <App />
      </ApolloProvider>
    )
    const query = wrapper.find(Query).prop('children')
    const queryWrapper = shallow(query({ loading: false, data: data, error: null }))

    expect(queryWrapper).toMatchSnapshot()
  })

  describe('<ComponentWrapper />', () => {
    it('renders correctly when loading', () => {
      const wrapper = shallow(<ComponentWrapper loading={true} />)
      expect(wrapper.text()).toEqual('Loading...')
    })

    it('renders correctly when erroring', () => {
      const error = {
        message: 'error message',
      }
      const wrapper = shallow(<ComponentWrapper loading={false} error={error} />)
      expect(wrapper.text()).toEqual(`Error! ${error.message}`)
    })

    it('renders correctly when data', () => {
      const wrapper = shallow(<ComponentWrapper loading={false} error={false} data={data} />)

      expect(wrapper.find(Profile).props()).toEqual(data.user)
      expect(wrapper.find(Repositories).prop('repositoriesList')).toEqual(data.user.repositories)
    })
  })
})
