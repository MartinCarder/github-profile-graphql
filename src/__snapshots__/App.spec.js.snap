// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should be the correct query 1`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "kind": "OperationDefinition",
      "name": undefined,
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": Object {
                  "kind": "Name",
                  "value": "login",
                },
                "value": Object {
                  "block": false,
                  "kind": "StringValue",
                  "value": "gaearon",
                },
              },
            ],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "user",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "avatarUrl",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "bio",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "company",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "name",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "location",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "login",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "first",
                      },
                      "value": Object {
                        "kind": "IntValue",
                        "value": "50",
                      },
                    },
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "orderBy",
                      },
                      "value": Object {
                        "fields": Array [
                          Object {
                            "kind": "ObjectField",
                            "name": Object {
                              "kind": "Name",
                              "value": "field",
                            },
                            "value": Object {
                              "kind": "EnumValue",
                              "value": "PUSHED_AT",
                            },
                          },
                          Object {
                            "kind": "ObjectField",
                            "name": Object {
                              "kind": "Name",
                              "value": "direction",
                            },
                            "value": Object {
                              "kind": "EnumValue",
                              "value": "DESC",
                            },
                          },
                        ],
                        "kind": "ObjectValue",
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "repositories",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "pageInfo",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "endCursor",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "startCursor",
                              },
                              "selectionSet": undefined,
                            },
                          ],
                        },
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "nodes",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "description",
                              },
                              "selectionSet": undefined,
                            },
                          ],
                        },
                      },
                    ],
                  },
                },
              ],
            },
          },
        ],
      },
      "variableDefinitions": Array [],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 357,
    "start": 0,
  },
}
`;

exports[`<App /> snapshot of App 1`] = `
<div
  className="App"
>
  <Profile
    avatarUrl="http://"
    login="martincarder"
    name="martin carder"
    repositories={
      Object {
        "nodes": Array [],
      }
    }
  />
  <Repositories
    repositoriesList={
      Object {
        "nodes": Array [],
      }
    }
  />
</div>
`;
